File: answers/other/answers/lintcode/Palindrome Partitioning II.cpp
class Solution {
public:
    /**
     * @param s a string
     * @return an integer
     */
    int minCut(string s) {
        // write your code here
        vector<int> min_cut(s.length(), INT_MAX);
        vector<vector<bool> > palin(s.length(), vector<bool> (s.length(), false));
        for (size_t i = 0; i < s.length(); ++ i) {
            min_cut[0] = 0;
            for (size_t j = 0; i + j < s.length(); ++ j) {
                if (0 == i) {
                    palin[i][j] = true;
                }
                else if (1 == i) {
                    palin[i][j] = s[j] == s[i + j];
                }
                else {
                    palin[i][j] = palin[i - 2][j + 1] && s[j] == s[i + j];
                }
            }
        }
        min_cut[0] = 0;
        for (size_t i = 0; i < s.length(); ++ i) {
            for (size_t j = i; j < s.length(); ++ j) {
                if (palin[j - i][i]) {
                    min_cut[j] = min(min_cut[j], (i > 0 ? min_cut[i - 1] + 1 : 0));
                }
            }
        }
        return min_cut[s.length() - 1];
    }
};

File: answers/leetcode/Palindrome Partitioning II/Palindrome Partitioning II.java
class Solution {
    public int minCut(String s) {
        int length = s.length();
        boolean[][] dp = new boolean[length][length];
        for (int i = 0; i < length; i++) {
            for (int j = 0; i + j < length; j++) {
                boolean flag = false;
                if (i == 0) {
                    flag = true;
                } else if (i == 1) {
                    flag = s.charAt(j) == s.charAt(i + j);
                } else {
                    flag = dp[j + 1][i + j - 1] && s.charAt(j) == s.charAt(i + j);
                }
                dp[j][i + j] = flag;
            }
        }
        int[] cut = new int[length + 1];
        for (int i = 1; i <= length; i++) {
            cut[i] = length + 1;
        }
        for (int i = 0; i < length; i++) {
            for (int j = i; j < length; j++) {
                if (dp[i][j]) {
                    cut[j + 1] = Math.min(cut[j + 1], cut[i] + 1);
                }
            }
        }
        return cut[length] - 1;
    }
}